File: Graph(그래프)/백준_17086_아기상어2/아기상어.py

from collections import deque

n,m = map(int, input().split(' '))
max_cnt = 0
q = deque()
mx = [-1, -1, -1, 0, 1, 1, 1, 0]
my = [-1, 0, 1, 1, 1, 0, -1, -1]
sea = [[0 for _ in range(51)] for _ in range(51)]
class shark:
    x = 0
    y = 0
    def __init__(self, x, y):
        self.x = x
        self.y = y

def bfs():
    global max_cnt     # 상수는 메서드안에서 사용하려면, 전역변수화 시켜야함
    while len(q) != 0:
        sh = q.popleft()
        for p in range(8):
            new_x = sh.x + mx[p]
            new_y = sh.y + my[p]
            if new_x >= 0 and new_x < n and new_y >= 0 and new_y < m:
                if sea[new_x][new_y] == 0:
                    sea[new_x][new_y] = sea[sh.x][sh.y] + 1
                    sh_possible = shark(new_x, new_y)
                    q.append(sh_possible)
                    if sea[new_x][new_y] > max_cnt:
                        max_cnt = sea[new_x][new_y]

for i in range(0, n):
    inpu = input().split(' ')
    for j in range(0,m):
        sea[i][j] = int(inpu[j])
        sh_position = shark(i, j)
        if sea[i][j] == 1: q.append(sh_position)

bfs()
print(max_cnt-1)

File: Graph(그래프)/백준_17086_아기상어2/아기상어.kt
import java.util.*

var n = 0
var m = 0
var max_cnt = 0

class shark {
    var x = 0
    var y = 0
    constructor(x: Int, y: Int) {
        this.x = x
        this.y = y
    }
}

var sea = Array(51, { Array(51, { 0 }) })
val mx = intArrayOf(-1, -1, -1, 0, 1, 1, 1, 0)
val my = intArrayOf(-1, 0, 1, 1, 1, 0, -1, -1)
var q = LinkedList<shark>()

fun BFS() {
    while (q.isEmpty() == false) {
        var sh = q.peek()
        q.poll()
        for (i in 0..7) {
            var new_x = sh.x + mx[i]
            var new_y = sh.y + my[i]
            if (new_x >= 0 && new_x < n && new_y >= 0 && new_y < m) {
                if (sea[new_x][new_y] == 0) {
                    sea[new_x][new_y] = sea[sh.x][sh.y] + 1
                    var sh_possible = shark(new_x, new_y)
                    q.add(sh_possible)
                    if (sea[new_x][new_y] > max_cnt) max_cnt = sea[new_x][new_y]
                }
            }
        }
    }
}

fun main(args: Array<String>) {
    var s = Scanner(System.`in`)
    n = s.nextInt()
    m = s.nextInt()
    for (i in 0..n - 1) {
        for (j in 0..m - 1) {
            var k = s.nextInt()
            sea[i][j] = k
            var sh_position = shark(i, j)
            if (k == 1) q.add(sh_position)
        }
    }
    BFS()

    print(max_cnt - 1)
}
